MaxText/layers/attentions.py

```diff
@@ -442,13 +284,48 @@ def apply_attention_dot(
     # Normalize the attention weights across `kv_length` dimension.
     attn_weights = jax.nn.softmax(attn_weights).astype(self.dtype)
 
-    # Apply attention dropout.
-    if not deterministic and self.dropout_rate > 0.:
-      attn_weights = self.attention_dropout(attn_weights, dropout_rng)
-
     # Take the linear combination of `value`.
     return self.wv_product(attn_weights, value, aqt_rng)
 
+  def qk_product(self, query: Array, key: Array) -> Array:
+    """Query-Key product.
+    
+    Args:
+      query: Query projection, in shape of [b, t, n, d], where b: batch size, t:
+        query length, n: number of heads, d: project dimension. 
+      key: Key projection in shape of [b, s, n_kv, d] for where n_kv is 
+        kv heads. The number of group for query is n // n_kv.
+
+    Returns:
+      results in shape [b, n_kv, n // n_kv,  t, s].
+    """
+    b, t, n, d = query.shape
+    n_kv = key.shape[-2]
```

Comment:
Ya I think it is a good question. Generally it is an attempt to make the code support variable number of fields for more exotic attentioons cases. I think it is OK to leave in for now (since it does have benefits) but I'm suspicious of it as well.